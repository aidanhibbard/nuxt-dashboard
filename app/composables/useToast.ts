import { ref, readonly } from 'vue'

export interface Toast {
  id: string
  type: 'success' | 'error' | 'warning' | 'info' | 'loading'
  message: string
  duration?: number
}

const toasts = ref<Toast[]>([])

export const useToast = () => {

  const addToast = (toast: Omit<Toast, 'id'>) => {
    const id = Date.now().toString()
    const newToast: Toast = {
      ...toast,
      id,
      duration: toast.duration ?? (toast.type === 'loading' ? 0 : 5000),
    }

    toasts.value.push(newToast)

    // Auto remove non-loading toasts
    if (newToast.duration && newToast.duration > 0) {
      setTimeout(() => {
        removeToast(id)
      }, newToast.duration)
    }

    return id
  }

  const removeToast = (id: string) => {
    const index = toasts.value.findIndex(toast => toast.id === id)
    if (index > -1) {
      toasts.value.splice(index, 1)
    }
  }

  const showSuccess = (message: string, duration?: number) => {
    return addToast({ type: 'success', message, duration })
  }

  const showError = (message: string, duration?: number) => {
    return addToast({ type: 'error', message, duration })
  }

  const showWarning = (message: string, duration?: number) => {
    return addToast({ type: 'warning', message, duration })
  }

  const showInfo = (message: string, duration?: number) => {
    return addToast({ type: 'info', message, duration })
  }

  const showLoading = (message: string) => {
    return addToast({ type: 'loading', message, duration: 0 })
  }

  const dismiss = (id: string) => {
    removeToast(id)
  }

  return {
    toasts: readonly(toasts),
    showSuccess,
    showError,
    showWarning,
    showInfo,
    showLoading,
    dismiss,
    removeToast,
  }
}
