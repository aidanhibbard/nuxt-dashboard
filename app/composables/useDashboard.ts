import { ref, computed, readonly } from 'vue'
import { useToast } from './useToast'

export interface DashboardStats {
  totalUsers: number
  activeUsers: number
  newUsers: number
  totalRevenue: number
  monthlyGrowth: number
  conversionRate: number
}

export interface RecentActivity {
  id: string
  type: 'user_created' | 'user_updated' | 'user_deleted' | 'login' | 'report_generated'
  message: string
  timestamp: Date
  user?: string
}

export interface ChartData {
  labels: string[]
  datasets: {
    label: string
    data: number[]
    backgroundColor?: string | string[]
    borderColor?: string | string[]
  }[]
}

export const useDashboard = () => {
  const { showSuccess, showError, showLoading, dismiss } = useToast()
  const loading = ref(false)
  const lastUpdated = ref(new Date())

  // Mock dashboard stats
  const stats = ref<DashboardStats>({
    totalUsers: 1247,
    activeUsers: 892,
    newUsers: 45,
    totalRevenue: 125000,
    monthlyGrowth: 12.5,
    conversionRate: 3.2,
  })

  // Mock recent activity
  const recentActivity = ref<RecentActivity[]>([
    {
      id: '1',
      type: 'user_created',
      message: 'New user registered',
      timestamp: new Date('2024-01-15T14:30:00Z'),
      user: 'john.doe@example.com',
    },
    {
      id: '2',
      type: 'login',
      message: 'User logged in',
      timestamp: new Date('2024-01-15T14:25:00Z'),
      user: 'jane.smith@example.com',
    },
    {
      id: '3',
      type: 'report_generated',
      message: 'Monthly report generated',
      timestamp: new Date('2024-01-15T14:20:00Z'),
      user: 'admin@example.com',
    },
    {
      id: '4',
      type: 'user_updated',
      message: 'User profile updated',
      timestamp: new Date('2024-01-15T14:15:00Z'),
      user: 'bob.johnson@example.com',
    },
    {
      id: '5',
      type: 'user_deleted',
      message: 'User account deleted',
      timestamp: new Date('2024-01-15T14:10:00Z'),
      user: 'inactive@example.com',
    },
  ])

  // Mock chart data
  const userGrowthData = ref<ChartData>({
    labels: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun'],
    datasets: [
      {
        label: 'New Users',
        data: [65, 78, 90, 85, 95, 120],
        backgroundColor: 'rgba(59, 130, 246, 0.2)',
        borderColor: 'rgba(59, 130, 246, 1)',
      },
    ],
  })

  const revenueData = ref<ChartData>({
    labels: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun'],
    datasets: [
      {
        label: 'Revenue',
        data: [12000, 15000, 18000, 16000, 22000, 25000],
        backgroundColor: 'rgba(34, 197, 94, 0.2)',
        borderColor: 'rgba(34, 197, 94, 1)',
      },
    ],
  })

  const userRoleDistribution = ref<ChartData>({
    labels: ['Admin', 'Manager', 'User', 'Viewer'],
    datasets: [
      {
        label: 'Users by Role',
        data: [5, 12, 45, 23],
        backgroundColor: [
          'rgba(239, 68, 68, 0.8)',
          'rgba(245, 158, 11, 0.8)',
          'rgba(59, 130, 246, 0.8)',
          'rgba(107, 114, 128, 0.8)',
        ],
      },
    ],
  })

  // Computed properties
  const formattedStats = computed(() => ({
    totalUsers: stats.value.totalUsers.toLocaleString(),
    activeUsers: stats.value.activeUsers.toLocaleString(),
    newUsers: stats.value.newUsers.toLocaleString(),
    totalRevenue: `$${stats.value.totalRevenue.toLocaleString()}`,
    monthlyGrowth: `${stats.value.monthlyGrowth}%`,
    conversionRate: `${stats.value.conversionRate}%`,
  }))

  const activityByType = computed(() => {
    return recentActivity.value.reduce((acc, activity) => {
      acc[activity.type] = (acc[activity.type] || 0) + 1
      return acc
    }, {} as Record<string, number>)
  })

  // Methods
  const refreshDashboard = async () => {
    const loadingToast = showLoading('Refreshing dashboard...')

    try {
      // Simulate API call
      await new Promise(resolve => setTimeout(resolve, 1500))

      // Update mock data with slight variations
      stats.value = {
        totalUsers: stats.value.totalUsers + Math.floor(Math.random() * 10),
        activeUsers: stats.value.activeUsers + Math.floor(Math.random() * 5),
        newUsers: Math.floor(Math.random() * 20) + 30,
        totalRevenue: stats.value.totalRevenue + Math.floor(Math.random() * 5000),
        monthlyGrowth: stats.value.monthlyGrowth + (Math.random() - 0.5) * 2,
        conversionRate: stats.value.conversionRate + (Math.random() - 0.5) * 0.5,
      }

      lastUpdated.value = new Date()
      dismiss(loadingToast)
      showSuccess('Dashboard refreshed successfully')
    }
    catch {
      dismiss(loadingToast)
      showError('Failed to refresh dashboard')
    }
  }

  const addActivity = (activity: Omit<RecentActivity, 'id'>) => {
    const newActivity: RecentActivity = {
      ...activity,
      id: Date.now().toString(),
    }
    recentActivity.value.unshift(newActivity)

    // Keep only last 10 activities
    if (recentActivity.value.length > 10) {
      recentActivity.value = recentActivity.value.slice(0, 10)
    }
  }

  return {
    stats: readonly(stats),
    formattedStats,
    recentActivity: readonly(recentActivity),
    activityByType,
    userGrowthData: readonly(userGrowthData),
    revenueData: readonly(revenueData),
    userRoleDistribution: readonly(userRoleDistribution),
    loading: readonly(loading),
    lastUpdated: readonly(lastUpdated),
    refreshDashboard,
    addActivity,
  }
}
