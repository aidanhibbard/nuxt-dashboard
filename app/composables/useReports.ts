import { ref, computed, readonly } from 'vue'
import { z } from 'zod'
import { useToast } from './useToast'

// Zod schemas
export const reportFiltersSchema = z.object({
  startDate: z.date(),
  endDate: z.date(),
  reportType: z.enum(['users', 'revenue', 'activity', 'performance']),
  groupBy: z.enum(['day', 'week', 'month']).optional(),
})

export type ReportFilters = z.infer<typeof reportFiltersSchema>

export interface ReportData {
  id: string
  type: string
  title: string
  data: unknown
  generatedAt: Date
  filters: ReportFilters
}

export interface ChartDataPoint {
  date: string
  value: number
  label?: string
}

export const useReports = () => {
  const { showSuccess, showError, showLoading, dismiss } = useToast()
  const loading = ref(false)
  const filters = ref<ReportFilters>({
    startDate: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000), // 30 days ago
    endDate: new Date(),
    reportType: 'users',
    groupBy: 'day',
  })

  // Mock report data
  const reports = ref<ReportData[]>([])
  const currentReport = ref<ReportData | null>(null)

  // Mock chart data generators
  const generateUserData = (startDate: Date, endDate: Date, groupBy: string = 'day') => {
    const data: ChartDataPoint[] = []
    const current = new Date(startDate)

    while (current <= endDate) {
      data.push({
        date: current.toISOString().split('T')[0]!,
        value: Math.floor(Math.random() * 100) + 50,
        label: current.toLocaleDateString(),
      })

      if (groupBy === 'day') {
        current.setDate(current.getDate() + 1)
      }
      else if (groupBy === 'week') {
        current.setDate(current.getDate() + 7)
      }
      else {
        current.setMonth(current.getMonth() + 1)
      }
    }

    return data
  }

  const generateRevenueData = (startDate: Date, endDate: Date, groupBy: string = 'day') => {
    const data: ChartDataPoint[] = []
    const current = new Date(startDate)

    while (current <= endDate) {
      data.push({
        date: current.toISOString().split('T')[0]!,
        value: Math.floor(Math.random() * 10000) + 5000,
        label: current.toLocaleDateString(),
      })

      if (groupBy === 'day') {
        current.setDate(current.getDate() + 1)
      }
      else if (groupBy === 'week') {
        current.setDate(current.getDate() + 7)
      }
      else {
        current.setMonth(current.getMonth() + 1)
      }
    }

    return data
  }

  const generateActivityData = (startDate: Date, endDate: Date, groupBy: string = 'day') => {
    const data: ChartDataPoint[] = []
    const current = new Date(startDate)

    while (current <= endDate) {
      data.push({
        date: current.toISOString().split('T')[0]!,
        value: Math.floor(Math.random() * 500) + 100,
        label: current.toLocaleDateString(),
      })

      if (groupBy === 'day') {
        current.setDate(current.getDate() + 1)
      }
      else if (groupBy === 'week') {
        current.setDate(current.getDate() + 7)
      }
      else {
        current.setMonth(current.getMonth() + 1)
      }
    }

    return data
  }

  // Computed properties
  const filteredReports = computed(() => {
    return reports.value.filter((report) => {
      const matchesType = (
        filters.value.reportType === undefined
        || filters.value.reportType === null
        || report.type === filters.value.reportType
      )
      const matchesDateRange = (
        report.generatedAt >= filters.value.startDate
        && report.generatedAt <= filters.value.endDate
      )
      return matchesType && matchesDateRange
    })
  })

  const reportSummary = computed(() => {
    if (!currentReport.value) return null

    const data = currentReport.value.data
    if (Array.isArray(data)) {
      const total = data.reduce((sum, item) => sum + item.value, 0)
      const average = total / data.length
      const max = Math.max(...data.map(item => item.value))
      const min = Math.min(...data.map(item => item.value))

      return {
        total,
        average: Math.round(average * 100) / 100,
        max,
        min,
        count: data.length,
      }
    }

    return null
  })

  // Methods
  const generateReport = async (reportFilters: ReportFilters) => {
    const loadingToast = showLoading('Generating report...')

    try {
      // Validate filters
      const validatedFilters = reportFiltersSchema.parse(reportFilters)

      // Simulate API call
      await new Promise(resolve => setTimeout(resolve, 2000))

      let reportData: unknown

      switch (validatedFilters.reportType) {
        case 'users':
          reportData = generateUserData(validatedFilters.startDate, validatedFilters.endDate, validatedFilters.groupBy)
          break
        case 'revenue':
          reportData = generateRevenueData(validatedFilters.startDate, validatedFilters.endDate, validatedFilters.groupBy)
          break
        case 'activity':
          reportData = generateActivityData(validatedFilters.startDate, validatedFilters.endDate, validatedFilters.groupBy)
          break
        case 'performance':
          reportData = {
            users: generateUserData(validatedFilters.startDate, validatedFilters.endDate, validatedFilters.groupBy),
            revenue: generateRevenueData(validatedFilters.startDate, validatedFilters.endDate, validatedFilters.groupBy),
          }
          break
      }

      const newReport: ReportData = {
        id: Date.now().toString(),
        type: validatedFilters.reportType,
        title: `${validatedFilters.reportType.charAt(0).toUpperCase() + validatedFilters.reportType.slice(1)} Report`,
        data: reportData,
        generatedAt: new Date(),
        filters: validatedFilters,
      }

      reports.value.unshift(newReport)
      currentReport.value = newReport
      filters.value = validatedFilters

      dismiss(loadingToast)
      showSuccess('Report generated successfully')

      return newReport
    }
    catch (error) {
      dismiss(loadingToast)
      if (error instanceof z.ZodError) {
        showError(error.errors[0]?.message ?? 'Validation error')
      }
      else {
        showError('Failed to generate report')
      }
      throw error
    }
  }

  const exportReport = async (reportId: string, format: 'csv' | 'json' | 'pdf' = 'csv') => {
    const loadingToast = showLoading(`Exporting report as ${format.toUpperCase()}...`)

    try {
      const report = reports.value.find(r => r.id === reportId)
      if (!report) {
        throw new Error('Report not found')
      }

      // Simulate export process
      await new Promise(resolve => setTimeout(resolve, 1500))

      dismiss(loadingToast)
      showSuccess(`Report exported as ${format.toUpperCase()} successfully`)

      // In a real app, this would trigger a download
      console.log(`Exporting report ${reportId} as ${format}`, report.data)
    }
    catch {
      dismiss(loadingToast)
      showError('Failed to export report')
    }
  }

  const deleteReport = async (reportId: string) => {
    const loadingToast = showLoading('Deleting report...')

    try {
      const reportIndex = reports.value.findIndex(r => r.id === reportId)
      if (reportIndex === -1) {
        throw new Error('Report not found')
      }

      // Simulate API call
      await new Promise(resolve => setTimeout(resolve, 800))

      reports.value.splice(reportIndex, 1)

      if (currentReport.value?.id === reportId) {
        currentReport.value = null
      }

      dismiss(loadingToast)
      showSuccess('Report deleted successfully')
    }
    catch {
      dismiss(loadingToast)
      showError('Failed to delete report')
    }
  }

  const setCurrentReport = (reportId: string) => {
    const report = reports.value.find(r => r.id === reportId)
    currentReport.value = report || null
  }

  return {
    filters,
    reports: readonly(reports),
    currentReport: readonly(currentReport),
    filteredReports,
    reportSummary,
    loading: readonly(loading),
    generateReport,
    exportReport,
    deleteReport,
    setCurrentReport,
  }
}
