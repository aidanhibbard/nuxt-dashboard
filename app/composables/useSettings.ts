import { ref, computed, readonly } from 'vue'
import { z } from 'zod'
import { useToast } from './useToast'

// Zod schemas
export const profileSchema = z.object({
  firstName: z.string().min(2, 'First name must be at least 2 characters'),
  lastName: z.string().min(2, 'Last name must be at least 2 characters'),
  email: z.string().email('Invalid email address'),
  avatar: z.string().url().optional(),
})

export const preferencesSchema = z.object({
  theme: z.enum(['light', 'dark', 'system']),
  language: z.enum(['en', 'es', 'fr', 'de']),
  timezone: z.string(),
  notifications: z.object({
    email: z.boolean(),
    push: z.boolean(),
    sms: z.boolean(),
  }),
  dashboard: z.object({
    defaultView: z.enum(['overview', 'analytics', 'reports']),
    refreshInterval: z.number().min(30).max(3600),
    showCharts: z.boolean(),
    showActivity: z.boolean(),
  }),
})

export const securitySchema = z.object({
  currentPassword: z.string().min(1, 'Current password is required'),
  newPassword: z.string().min(8, 'Password must be at least 8 characters'),
  confirmPassword: z.string().min(1, 'Please confirm your password'),
}).refine(data => data.newPassword === data.confirmPassword, {
  message: 'Passwords don\'t match',
  path: ['confirmPassword'],
})

export type Profile = z.infer<typeof profileSchema>
export type Preferences = z.infer<typeof preferencesSchema>
export type SecuritySettings = z.infer<typeof securitySchema>

export const useSettings = () => {
  const { showSuccess, showError, showLoading, dismiss } = useToast()
  const loading = ref(false)

  // Mock user profile
  const profile = ref<Profile>({
    firstName: 'John',
    lastName: 'Doe',
    email: 'john.doe@example.com',
    avatar: 'https://images.unsplash.com/photo-1472099645785-5658abf4ff4e?w=150&h=150&fit=crop&crop=face',
  })

  // Mock user preferences
  const preferences = ref<Preferences>({
    theme: 'system',
    language: 'en',
    timezone: 'America/New_York',
    notifications: {
      email: true,
      push: true,
      sms: false,
    },
    dashboard: {
      defaultView: 'overview',
      refreshInterval: 300,
      showCharts: true,
      showActivity: true,
    },
  })

  // Computed properties
  const fullName = computed(() => {
    return `${profile.value.firstName} ${profile.value.lastName}`
  })

  const initials = computed(() => {
    return `${profile.value.firstName.charAt(0)}${profile.value.lastName.charAt(0)}`
  })

  const isDarkMode = computed(() => {
    if (preferences.value.theme === 'system') {
      return window.matchMedia('(prefers-color-scheme: dark)').matches
    }
    return preferences.value.theme === 'dark'
  })

  // Methods
  const updateProfile = async (profileData: Partial<Profile>) => {
    const loadingToast = showLoading('Updating profile...')

    try {
      // Validate data
      const validatedData = profileSchema.partial().parse(profileData)

      // Simulate API call
      await new Promise(resolve => setTimeout(resolve, 1000))

      Object.assign(profile.value, validatedData)

      dismiss(loadingToast)
      showSuccess('Profile updated successfully')

      return profile.value
    }
    catch (error) {
      dismiss(loadingToast)
      if (error instanceof z.ZodError) {
        showError(error.errors[0]?.message ?? 'Failed to update profile')
      }
      else {
        showError('Failed to update profile')
      }
      throw error
    }
  }

  const updatePreferences = async (preferencesData: Partial<Preferences>) => {
    const loadingToast = showLoading('Updating preferences...')

    try {
      // Validate data
      const validatedData = preferencesSchema.partial().parse(preferencesData)

      // Simulate API call
      await new Promise(resolve => setTimeout(resolve, 800))

      Object.assign(preferences.value, validatedData)

      // Apply theme changes immediately
      if (validatedData.theme) {
        applyTheme(validatedData.theme)
      }

      dismiss(loadingToast)
      showSuccess('Preferences updated successfully')

      return preferences.value
    }
    catch (error) {
      dismiss(loadingToast)
      if (error instanceof z.ZodError) {
        showError(error.errors[0]?.message ?? 'Validation error')
      }
      else {
        showError('Failed to update preferences')
      }
      throw error
    }
  }

  const changePassword = async (securityData: SecuritySettings) => {
    const loadingToast = showLoading('Changing password...')

    try {
      // Validate data
      const validatedData = securitySchema.parse(securityData)

      // Simulate API call
      await new Promise(resolve => setTimeout(resolve, 1200))

      dismiss(loadingToast)
      showSuccess('Password changed successfully')

      // In a real app, you would update the user's password here
      console.log('Password changed:', validatedData.newPassword)
    }
    catch (error) {
      dismiss(loadingToast)
      if (error instanceof z.ZodError) {
        showError(error.errors[0]?.message ?? 'Failed to change password')
      }
      else {
        showError('Failed to change password')
      }
      throw error
    }
  }

  const uploadAvatar = async (file: File) => {
    const loadingToast = showLoading('Uploading avatar...')

    try {
      // Validate file
      if (!file.type.startsWith('image/')) {
        throw new Error('Please select an image file')
      }

      if (file.size > 5 * 1024 * 1024) { // 5MB limit
        throw new Error('File size must be less than 5MB')
      }

      // Simulate upload
      await new Promise(resolve => setTimeout(resolve, 1500))

      // In a real app, you would upload to a server and get back a URL
      const mockAvatarUrl = `https://images.unsplash.com/photo-${Date.now()}?w=150&h=150&fit=crop&crop=face`

      profile.value.avatar = mockAvatarUrl

      dismiss(loadingToast)
      showSuccess('Avatar uploaded successfully')

      return mockAvatarUrl
    }
    catch (error) {
      dismiss(loadingToast)
      showError(error instanceof Error ? error.message : 'Failed to upload avatar')
      throw error
    }
  }

  const applyTheme = (theme: 'light' | 'dark' | 'system') => {
    // In a real app, you would apply the theme to the document
    if (theme === 'dark' || (theme === 'system' && window.matchMedia('(prefers-color-scheme: dark)').matches)) {
      document.documentElement.classList.add('dark')
    }
    else {
      document.documentElement.classList.remove('dark')
    }
  }

  const exportData = async () => {
    const loadingToast = showLoading('Exporting data...')

    try {
      // Simulate export process
      await new Promise(resolve => setTimeout(resolve, 2000))

      const exportData = {
        profile: profile.value,
        preferences: preferences.value,
        exportedAt: new Date().toISOString(),
      }

      dismiss(loadingToast)
      showSuccess('Data exported successfully')

      // In a real app, this would trigger a download
      console.log('Exporting data:', exportData)

      return exportData
    }
    catch (error) {
      dismiss(loadingToast)
      showError('Failed to export data')
      throw error
    }
  }

  const deleteAccount = async (confirmation: string) => {
    if (confirmation !== 'DELETE') {
      showError('Please type DELETE to confirm account deletion')
      return
    }

    const loadingToast = showLoading('Deleting account...')

    try {
      // Simulate API call
      await new Promise(resolve => setTimeout(resolve, 3000))

      dismiss(loadingToast)
      showSuccess('Account deleted successfully')

      // In a real app, you would redirect to logout or home page
      console.log('Account deleted')
    }
    catch (error) {
      dismiss(loadingToast)
      showError('Failed to delete account')
      throw error
    }
  }

  return {
    profile: readonly(profile),
    preferences: readonly(preferences),
    fullName,
    initials,
    isDarkMode,
    loading: readonly(loading),
    updateProfile,
    updatePreferences,
    changePassword,
    uploadAvatar,
    applyTheme,
    exportData,
    deleteAccount,
  }
}
