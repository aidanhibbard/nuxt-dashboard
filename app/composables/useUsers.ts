import { ref, computed, readonly } from 'vue'
import { z } from 'zod'
import { useToast } from './useToast'

// Zod schemas
export const userSchema = z.object({
  id: z.string().optional(),
  name: z.string().min(2, 'Name must be at least 2 characters'),
  email: z.string().email('Invalid email address'),
  role: z.enum(['admin', 'manager', 'user', 'viewer']),
  status: z.enum(['active', 'inactive', 'pending']),
  avatar: z.string().url().optional(),
  lastLogin: z.date().optional(),
})

export type User = z.infer<typeof userSchema>

// Mock data
const mockUsers: User[] = [
  {
    id: '1',
    name: 'John Doe',
    email: 'john.doe@example.com',
    role: 'admin',
    status: 'active',
    avatar: 'https://images.unsplash.com/photo-1472099645785-5658abf4ff4e?w=150&h=150&fit=crop&crop=face',
    lastLogin: new Date('2024-01-15T10:30:00Z'),
  },
  {
    id: '2',
    name: 'Jane Smith',
    email: 'jane.smith@example.com',
    role: 'manager',
    status: 'active',
    avatar: 'https://images.unsplash.com/photo-1494790108755-2616b612b786?w=150&h=150&fit=crop&crop=face',
    lastLogin: new Date('2024-01-14T15:45:00Z'),
  },
  {
    id: '3',
    name: 'Bob Johnson',
    email: 'bob.johnson@example.com',
    role: 'user',
    status: 'active',
    avatar: 'https://images.unsplash.com/photo-1507003211169-0a1dd7228f2d?w=150&h=150&fit=crop&crop=face',
    lastLogin: new Date('2024-01-13T09:20:00Z'),
  },
  {
    id: '4',
    name: 'Alice Brown',
    email: 'alice.brown@example.com',
    role: 'viewer',
    status: 'inactive',
    avatar: 'https://images.unsplash.com/photo-1438761681033-6461ffad8d80?w=150&h=150&fit=crop&crop=face',
    lastLogin: new Date('2024-01-10T14:15:00Z'),
  },
  {
    id: '5',
    name: 'Charlie Wilson',
    email: 'charlie.wilson@example.com',
    role: 'user',
    status: 'pending',
    avatar: 'https://images.unsplash.com/photo-1500648767791-00dcc994a43e?w=150&h=150&fit=crop&crop=face',
    lastLogin: undefined,
  },
]

export const useUsers = () => {
  const { showSuccess, showError, showLoading, dismiss } = useToast()
  const users = ref<User[]>(mockUsers)
  const loading = ref(false)
  const searchQuery = ref('')
  const statusFilter = ref<string>('all')
  const roleFilter = ref<string>('all')

  // Computed properties
  const filteredUsers = computed(() => {
    return users.value.filter((user) => {
      const matchesSearch = user.name.toLowerCase().includes(searchQuery.value.toLowerCase())
        || user.email.toLowerCase().includes(searchQuery.value.toLowerCase())
      const matchesStatus = statusFilter.value === 'all' || user.status === statusFilter.value
      const matchesRole = roleFilter.value === 'all' || user.role === roleFilter.value

      return matchesSearch && matchesStatus && matchesRole
    })
  })

  const usersByRole = computed(() => {
    const grouped = users.value.reduce((acc, user) => {
      acc[user.role] = (acc[user.role] || 0) + 1
      return acc
    }, {} as Record<string, number>)
    return grouped
  })

  // CRUD operations
  const createUser = async (userData: Omit<User, 'id'>) => {
    const loadingToast = showLoading('Creating user...')

    try {
      // Validate data
      const validatedData = userSchema.parse(userData)

      // Simulate API call
      await new Promise(resolve => setTimeout(resolve, 1000))

      const newUser: User = {
        ...validatedData,
        id: Date.now().toString(),
        lastLogin: undefined,
      }

      users.value.push(newUser)
      dismiss(loadingToast)
      showSuccess('User created successfully')

      return newUser
    }
    catch (error) {
      dismiss(loadingToast)
      showError('Failed to create user')
      throw error
    }
  }

  const updateUser = async (id: string, userData: Partial<User>) => {
    const loadingToast = showLoading('Updating user...')

    try {
      // Validate data
      const validatedData = userSchema.partial().parse(userData)

      // Simulate API call
      await new Promise(resolve => setTimeout(resolve, 800))

      const userIndex = users.value.findIndex(user => user.id === id)
      if (userIndex === -1) {
        throw new Error('User not found')
      }

      users.value[userIndex] = { ...users.value[userIndex], ...validatedData } as User
      dismiss(loadingToast)
      showSuccess('User updated successfully')

      return users.value[userIndex]
    }
    catch (error) {
      dismiss(loadingToast)
      showError('Failed to update user')
      throw error
    }
  }

  const deleteUser = async (id: string) => {
    const loadingToast = showLoading('Deleting user...')

    try {
      // Simulate API call
      await new Promise(resolve => setTimeout(resolve, 600))

      const userIndex = users.value.findIndex(user => user.id === id)
      if (userIndex === -1) {
        throw new Error('User not found')
      }

      users.value.splice(userIndex, 1)
      dismiss(loadingToast)
      showSuccess('User deleted successfully')
    }
    catch (error) {
      dismiss(loadingToast)
      showError('Failed to delete user')
      throw error
    }
  }

  const getUserById = (id: string) => {
    return users.value.find(user => user.id === id)
  }

  return {
    users: readonly(users),
    filteredUsers,
    usersByRole,
    loading: readonly(loading),
    searchQuery,
    statusFilter,
    roleFilter,
    createUser,
    updateUser,
    deleteUser,
    getUserById,
  }
}
