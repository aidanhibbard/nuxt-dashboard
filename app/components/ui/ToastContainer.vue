<script setup lang="ts">
import { useToast } from '@/composables/useToast'
import { Alert, AlertDescription } from '@/components/shadcn/ui/alert'
import { Button } from '@/components/shadcn/ui/button'
import { X, CheckCircle, AlertCircle, AlertTriangle, Info, Loader2 } from 'lucide-vue-next'

const { toasts, removeToast } = useToast()

const getToastIcon = (type: string) => {
  switch (type) {
    case 'success':
      return CheckCircle
    case 'error':
      return AlertCircle
    case 'warning':
      return AlertTriangle
    case 'info':
      return Info
    case 'loading':
      return Loader2
    default:
      return Info
  }
}

const getToastVariant = (type: string) => {
  switch (type) {
    case 'success':
      return 'default'
    case 'error':
      return 'destructive'
    case 'warning':
      return 'default'
    case 'info':
      return 'default'
    case 'loading':
      return 'default'
    default:
      return 'default'
  }
}
</script>

<template>
  <div class="fixed top-4 right-4 z-50 space-y-2">
    <TransitionGroup
      name="toast"
      tag="div"
      class="space-y-2"
    >
      <Alert
        v-for="toast in toasts"
        :key="toast.id"
        :variant="getToastVariant(toast.type)"
        class="w-80 shadow-lg border"
      >
        <div class="flex items-start gap-3">
          <component
            :is="getToastIcon(toast.type)"
            class="h-4 w-4 mt-0.5"
            :class="{
              'text-green-600': toast.type === 'success',
              'text-red-600': toast.type === 'error',
              'text-yellow-600': toast.type === 'warning',
              'text-blue-600': toast.type === 'info',
              'text-gray-600 animate-spin': toast.type === 'loading',
            }"
          />
          <AlertDescription class="flex-1">
            {{ toast.message }}
          </AlertDescription>
          <Button
            v-if="toast.type !== 'loading'"
            variant="ghost"
            size="sm"
            class="h-6 w-6 p-0"
            @click="removeToast(toast.id)"
          >
            <X class="h-3 w-3" />
          </Button>
        </div>
      </Alert>
    </TransitionGroup>
  </div>
</template>

<style scoped>
.toast-enter-active,
.toast-leave-active {
  transition: all 0.3s ease;
}

.toast-enter-from {
  opacity: 0;
  transform: translateX(100%);
}

.toast-leave-to {
  opacity: 0;
  transform: translateX(100%);
}

.toast-move {
  transition: transform 0.3s ease;
}
</style>
