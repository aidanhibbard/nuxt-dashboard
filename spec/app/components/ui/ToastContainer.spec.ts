import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import ToastContainer from '@/components/ui/ToastContainer.vue'
import { useToast } from '@/composables/useToast'

describe('components/ui/ToastContainer.vue', () => {
  it('renders and dismisses toasts', async () => {
    const wrapper = mount(ToastContainer)
    const { showSuccess, removeToast } = useToast()
    const id = showSuccess('Saved successfully', 0)
    await wrapper.vm.$nextTick()
    // mount a second tick to allow reactive flush
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toContain('Saved successfully')
    removeToast(id)
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).not.toContain('Saved successfully')
  })

  it('matches snapshot', () => {
    const wrapper = mount(ToastContainer)
    expect(wrapper.html()).toMatchSnapshot()
  })
})


