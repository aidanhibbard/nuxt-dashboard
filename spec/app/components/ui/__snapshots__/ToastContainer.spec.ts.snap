// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/ui/ToastContainer.vue > matches snapshot 1`] = `
"<div data-v-b0807160="" class="fixed top-4 right-4 z-50 space-y-2">
  <transition-group-stub data-v-b0807160="" name="toast" tag="div" appear="false" persisted="false" css="true" class="space-y-2"></transition-group-stub>
</div>"
`;
